### PR TITLE
sys_fs, sys_ss, and sys_usbd: LV2 syscalls enhancements

### PR DESCRIPTION
**Pull request by brian218**
- Implemented sys_fs_mount() with argument prot=1 for mounting VFS devices in read only mode (mainly used by XMB).
 - Further implemented Update Manager service so that the progress of firmware update in XMB can go to 99%, tested in DEX&DECR.
 - Implemented sys_usbd_unregister_extra_ldd() and improved the other ldd-related syscalls.

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_fs.h [hunk 1]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "Emu/IdManager.h"
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/lv2/sys_fs.h:3]: don't add this here.
You can just move the constructor+member init to the cpp...

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.h]: Done. Thanks for advice.

File: rpcs3/Emu/Cell/lv2/sys_ss.cpp [hunk 2]
```diff
@@ -300,18 +311,40 @@ error_code sys_ss_update_manager(u64 pkg_id, u64 a1, u64 a2, u64 a3, u64 a4, u64
...
+		const auto addr = vm::alloc(a1, vm::main);
+		if (!addr)
+			return CELL_ENOMEM;
+		vm::write32(a2, addr);
+		break;
+	}
+	case 0x600F:
+	{
+		// release buffer
+		if (!vm::dealloc(a2, vm::main))
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_ss.cpp:343]: Never expose vm::dealloc directly, you must ensure that the origin of the allocation is from sys_ss.

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_ss.cpp]: Thanks for notifying.
Fixed it up.

**Review by elad335 — CHANGES_REQUESTED** [review]
Thread-safety matters.

**Comment by brian218** [comment]
> Thread-safety matters.

Pardon, did you refer to sys_fs or sys_ss of this PR?

**Comment by brian218** [comment]
Locks have been added to sys_fs_newfs/mount/unmount syscalls earlier today.

**Comment by elad335** [comment]
cellSysCache and sys_ss contain such code.

**Comment by brian218** [comment]
> cellSysCache and sys_ss contain such code.

Solved.

File: rpcs3/Emu/Cell/Modules/cellSysCache.cpp [hunk 3]
```diff
@@ -121,9 +123,12 @@ error_code cellSysCacheMount(vm::ptr<CellSysCacheParam> param)
...
 
 	if (!param || (param->cacheId[0] && sysutil_check_name_string(param->cacheId, 1, CELL_SYSCACHE_ID_SIZE) != 0))
-	{
 		return CELL_SYSCACHE_ERROR_PARAM;
-	}
+
+	std::unique_lock lock_hdd1(g_mp_sys_dev_hdd1.mutex, std::try_to_lock);
+
+	if (!lock_hdd1.owns_lock())
+		return CELL_EBUSY;
```
Comment (elad335) [inline rpcs3/Emu/Cell/Modules/cellSysCache.cpp:131]: Please do not reinvent PS3 operating system, the function is the same and this error code does not exist for it.

Comment (brian218) [reply rpcs3/Emu/Cell/Modules/cellSysCache.cpp]: Thanks for correcting.
I reverted most of that.

**Comment by brian218** [comment]
> Thread-safety matters.

A shared_mutex has been added to `lv2_update_manager` as per the request.

**Comment by brian218** [comment]
While waiting for this PR's getting attention from the major developers, I made the progress of XMB's firmware update feature go farther, though it hasn't been fully supported yet.

screenshot_2023_05_19_00_44_51
screenshot_2023_05_19_00_44_54
screenshot_2023_05_19_00_44_58
screenshot_2023_05_19_00_45_30

**Comment by Megamouse** [comment]
"While waiting for this PR's getting attention from the major developers"
It's not gonna get faster if you add more stuff xd

**Comment by brian218** [comment]
> "While waiting for this PR's getting attention from the major developers" It's not gonna get faster if you add more stuff xd

With new features and improvements, there's no loss for me honestly. XD

Comparison with `v0.0.27-15025-08c1a14b`:
screenshot_2023_05_19

**Comment by Megamouse** [comment]
hi, could you be so kind and test the usio config in #13896 ?

File: rpcs3/Emu/Cell/lv2/sys_ss.cpp [hunk 4]
```diff
@@ -128,7 +203,7 @@ error_code sys_ss_get_console_id(vm::ptr<u8> buf)
 {
 	sys_ss.notice("sys_ss_get_console_id(buf=*0x%x)", buf);
 
-	return sys_ss_appliance_info_manager(0x19003, buf);
+	return not_an_error(sys_ss_appliance_info_manager(0x19003, buf)); // not_an_error: Prevent duplicate error reports in emulator logs
```
Comment (elad335) [inline rpcs3/Emu/Cell/lv2/sys_ss.cpp:206]: No duplicate error report, error_code::error_code(error_code&&) is a trivial constructor.

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_ss.cpp]: Oh, I see. I didn't notice that.
Thank you very much!

**Review by elad335 — APPROVED** [review]

**Comment by Hadrianneue** [comment]
Gran Turismo 6 crashing on startup, works fine on previous build 0.0.28-15120.

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/11654965/RPCS3.zip)

**Comment by brian218** [comment]
> Gran Turismo 6 crashing on startup, works fine on previous build 0.0.28-15120.
> 
> [RPCS3.zip](https://github.com/RPCS3/rpcs3/files/11654965/RPCS3.zip)

Kindly submit a formal issue report at https://github.com/RPCS3/rpcs3/issues/new/choose with the required information.

File: Utilities/File.cpp [hunk 5]
```diff
@@ -384,55 +385,26 @@ shared_ptr<fs::device_base> fs::set_virtual_device(const std::string& name, shar
...
-	std::string_view result = path;
+	std::string normalized_path = std::filesystem::path(path).lexically_normal().string();
 
-	// Number of path components to remove
-	usz to_remove = levels;
+#ifdef _WIN32
+	std::replace(normalized_path.begin(), normalized_path.end(), '\\', '/');
+#endif
+
+	if (normalized_path.back() == '/')
```
[the remark this answers is not in the document]
Comment (brian218) [reply Utilities/File.cpp:396]: Thank you so much for pointing that out.
Please check out #13955 for the new fix.

File: rpcs3/Emu/Cell/lv2/sys_fs.cpp [hunk 6]
```diff
@@ -209,60 +220,51 @@ u64 lv2_fs_mount_info_map::get_all(CellFsMountInfo* info, u64 len) const
...
+	const auto result = vfs::unmount(vpath);
 
-		// Process special directories
-		if (name == "."sv)
-		{
-			// Keep current
-			continue;
-		}
+	if (result)
+		g_fxo->get<lv2_fs_mount_info_map>().remove(vpath);
```
Comment (oltolm) [inline rpcs3/Emu/Cell/lv2/sys_fs.cpp:245]: This code leads to a crash. You are removing from the map while iterating. The loop is on line 147.

Comment (Megamouse) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:245]: @oltolm can you try with #13961 ?

Comment (oltolm) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:245]: I have, it doesn't crash anymore.

Comment (brian218) [reply rpcs3/Emu/Cell/lv2/sys_fs.cpp:245]: Thank you two for working on that.
I feel sorry about that as the problems didn't occur when I was testing.

File: rpcs3/Emu/System.cpp [hunk 7]
```diff
@@ -3460,7 +3460,7 @@ utils::serial* Emulator::DeserialManager() const
 
 bool Emulator::IsVsh()
 {
-	return g_ps3_process_info.get_cellos_appname() == "vsh.self"sv;
+	return g_ps3_process_info.self_info.prog_id_hdr.program_authority_id >> 52 == 0x107; // Not only VSH but also most CoreOS LV2 SELFs need the special treatment
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:3463]: This getter was meant for vsh.self (as you can probably guess from the name).
I don't see how this random unexplained magic number combination can possibly fit all applications of IsVsh() in the codebase.
If anything this is confusing to say the least.

Comment (brian218) [reply rpcs3/Emu/System.cpp:3463]: During my tests with some CoreOS LV2 SELFs (like ps3swu.self and emer_init.self), mostly they turn out to need the same special treatment as VSH needs, and those SELFs all have the PAID starting with 0x107xxxxxxxxxxxxx.
Reference: https://www.psdevwiki.com/ps3/Program_Authority_ID#Structure

**Pull request merged**